File: packages/elixir/lumis/test/conformance_test.exs
defmodule Lumis.ConformanceTest do
  use ExUnit.Case, async: true
  @moduletag :conformance

  @conformance_dir Path.expand("../../../../fixtures/conformance", __DIR__)

  @fixture_names @conformance_dir
                 |> Path.join("*/fixture.json")
                 |> Path.wildcard()
                 |> Enum.map(&Path.dirname/1)
                 |> Enum.map(&Path.basename/1)
                 |> Enum.sort()

  # The same eleven the JavaScript conformance suite declares. Highlighting would
  # load them anyway; naming them keeps this suite comparing formatter output
  # rather than measuring a first load, and matches what the browser has to do,
  # since it cannot load inside the walk.
  setup_all do
    :ok =
      Lumis.Languages.load(~w(
        json diff elixir html javascript css lua markdown markdown_inline mdx python
      ))

    :ok
  end

  defp load_fixture(name) do
    dir = Path.join(@conformance_dir, name)
    metadata = dir |> Path.join("fixture.json") |> File.read!() |> Jason.decode!()
    metadata = Map.put(metadata, "htmlMultiThemesOptions", metadata["htmlMultiThemes"])

    Map.merge(metadata, %{
      "source" => File.read!(Path.join(dir, "source.txt")),
      "htmlInline" => File.read!(Path.join(dir, "html-inline.html")),
      "htmlLinked" => File.read!(Path.join(dir, "html-linked.html")),
      "htmlMultiThemes" => File.read!(Path.join(dir, "html-multi-themes.html")),
      "bbcode" => File.read!(Path.join(dir, "bbcode.txt")),
      "terminal" => File.read!(Path.join(dir, "terminal.txt"))
    })
  end

  defp rainbow_brackets(fixture), do: fixture["rainbowBrackets"] || false

  defp serialize_event({:start, %{scope: scope, language: language}}),
    do: %{"type" => "start", "scope" => scope, "language" => language}

  defp serialize_event({:source, %{start: start, end: stop}}),
    do: %{"type" => "source", "start" => start, "end" => stop}

  defp serialize_event(:end), do: %{"type" => "end"}

  defp serialize_event({:decoration_start, %Lumis.Decoration.RainbowBracket{depth: depth}}),
    do: %{
      "type" => "decorationStart",
      "decoration" => %{"type" => "rainbowBracket", "depth" => depth}
    }

  defp serialize_event(:decoration_end), do: %{"type" => "decorationEnd"}

  defp conformance_theme(name) do
    path = Path.expand("../../../../fixtures/conformance-themes/#{name}.json", __DIR__)

    case File.read(path) do
      {:ok, json} ->
        {:ok, theme} = Lumis.Theme.from_json(json)
        theme

      {:error, :enoent} ->
        name
    end
  end

  defp html_multi_themes_options(fixture) do
    config =
      fixture["htmlMultiThemesOptions"] ||
        %{
          "themes" => %{"main" => fixture["theme"]},
          "defaultTheme" => "main"
        }

    # Reverse-sorted on purpose: the formatter sorts theme names itself, so
    # output must not depend on the order they were given in.
    themes =
      config["themes"]
      |> Enum.sort(:desc)
      |> Enum.map(fn {name, theme} -> {String.to_atom(name), conformance_theme(theme)} end)

    highlight_lines =
      case config["highlightLines"] do
        lines when is_list(lines) and lines != [] -> %{lines: lines, style: :theme}
        _ -> nil
      end

    [
      language: fixture["language"],
      themes: themes,
      default_theme: config["defaultTheme"],
      highlight_lines: highlight_lines
    ]
  end

  for name <- @fixture_names do
    describe name do
      @tag fixture: name
      test "events" do
        fixture = load_fixture(unquote(name))

        events =
          Lumis.highlight_events!(fixture["source"], fixture["language"],
            rainbow_brackets: rainbow_brackets(fixture)
          )

        assert Enum.map(events, &serialize_event/1) == fixture["events"]
      end

      @tag fixture: name
      test "html_inline" do
        fixture = load_fixture(unquote(name))

        assert Lumis.highlight!(fixture["source"],
                 formatter: {
                   :html_inline,
                   language: fixture["language"], theme: fixture["theme"]
                 },
                 rainbow_brackets: rainbow_brackets(fixture)
               ) == fixture["htmlInline"]
      end

      @tag fixture: name
      test "html_linked" do
        fixture = load_fixture(unquote(name))

        assert Lumis.highlight!(fixture["source"],
                 formatter: {:html_linked, language: fixture["language"]},
                 rainbow_brackets: rainbow_brackets(fixture)
               ) == fixture["htmlLinked"]
      end

      @tag fixture: name
      test "html_multi_themes" do
        fixture = load_fixture(unquote(name))

        assert Lumis.highlight!(fixture["source"],
                 formatter: {
                   :html_multi_themes,
                   html_multi_themes_options(fixture)
                 },
                 rainbow_brackets: rainbow_brackets(fixture)
               ) == fixture["htmlMultiThemes"]
      end

      @tag fixture: name
      test "terminal" do
        fixture = load_fixture(unquote(name))

        assert Lumis.highlight!(fixture["source"],
                 formatter: {
                   :terminal,
                   language: fixture["language"], theme: fixture["theme"]
                 },
                 rainbow_brackets: rainbow_brackets(fixture)
               ) == fixture["terminal"]
      end

      @tag fixture: name
      test "bbcode_scoped" do
        fixture = load_fixture(unquote(name))

        assert Lumis.highlight!(fixture["source"],
                 formatter: {:bbcode_scoped, language: fixture["language"]},
                 rainbow_brackets: rainbow_brackets(fixture)
               ) == fixture["bbcode"]
      end
    end
  end
end

File: packages/javascript/lumis/test/conformance.test.ts
import { beforeAll, describe, expect, it } from "vitest";

import dracula from "../../themes/dist/json/dracula.json";
import dark from "../../../../fixtures/conformance-themes/dark.json";
import light from "../../../../fixtures/conformance-themes/light.json";
import mid from "../../../../fixtures/conformance-themes/mid.json";
import css from "../langs/css.ts";
import diff from "../langs/diff.ts";
import elixir from "../langs/elixir.ts";
import html from "../langs/html.ts";
import javascript from "../langs/javascript.ts";
import json from "../langs/json.ts";
import lua from "../langs/lua.ts";
import markdown from "../langs/markdown.ts";
import markdownInline from "../langs/markdown_inline.ts";
import mdx from "../langs/mdx.ts";
import python from "../langs/python.ts";
import rust from "../langs/rust.ts";
import { createHighlighter } from "../src/index.js";
import {
  bbcodeScoped,
  htmlInline,
  htmlLinked,
  htmlMultiThemes,
  terminal,
} from "../src/formatters.js";
import type { Highlighter } from "../src/core/highlighter.js";
import type { Language, Theme } from "../src/types.js";

import { loadConformanceFixtures } from "./conformance.js";
import { configureLocalWasmResolver } from "./wasm.js";

const conformanceFixtures = loadConformanceFixtures();
const themes: Record<string, Theme> = {
  dark,
  dracula,
  light,
  mid,
};

const langBundles: Record<string, Language> = {
  json,
  diff,
  elixir,
  html,
  javascript,
  css,
  lua,
  markdown,
  markdownInline,
  mdx,
  python,
  rust,
};

function getLanguage(id: string): Language {
  const language = langBundles[id];
  if (!language) throw new Error(`No bundle for language "${id}" in test setup`);
  return language;
}

function getTheme(id: string): Theme {
  const theme = themes[id];
  if (!theme) throw new Error(`No theme for "${id}" in test setup`);
  return theme;
}

let highlighter: Highlighter;

beforeAll(async () => {
  configureLocalWasmResolver([
    "diff",
    "elixir",
    "json",
    "html",
    "javascript",
    "css",
    "lua",
    "markdown",
    "markdown_inline",
    "python",
    "rust",
  ]);
  // No `wasm` overrides. Parser bytes and the package that describes them have
  // to come from the same place, or the integrity check rejects the pair: CI
  // stages freshly built parsers while `ensureLocalWasm` returns the committed
  // fixture, which is a different build of the same grammar.
  highlighter = await createHighlighter({
    languages: [
      json,
      diff,
      elixir,
      html,
      javascript,
      css,
      lua,
      markdown,
      markdownInline,
      mdx,
      python,
      rust,
    ],
  });
}, 120_000);

describe.each(conformanceFixtures.map((f) => [f.name, f]))("%s", (_name, fixture) => {
  it("html-inline", () => {
    const output = highlighter.highlight(
      fixture.source,
      htmlInline({
        language: getLanguage(fixture.language),
        theme: getTheme(fixture.theme),
      }),
      { rainbowBrackets: fixture.rainbowBrackets },
    );
    expect(output).toBe(fixture.htmlInline);
  });

  it("html-linked", () => {
    const output = highlighter.highlight(
      fixture.source,
      htmlLinked({
        language: getLanguage(fixture.language),
      }),
      { rainbowBrackets: fixture.rainbowBrackets },
    );
    expect(output).toBe(fixture.htmlLinked);
  });

  it("html-multi-themes", () => {
    const config = fixture.htmlMultiThemesOptions;
    // Reversed on purpose. Rust holds themes in a `HashMap` and sorts before
    // emitting, so output must not depend on the order the caller inserted
    // them. Feeding them in already-sorted order would not prove that.
    const formatterThemes = config
      ? Object.fromEntries(
          Object.entries(config.themes)
            .toReversed()
            .map(([name, theme]) => [name, getTheme(theme)]),
        )
      : { main: getTheme(fixture.theme) };
    const output = highlighter.highlight(
      fixture.source,
      htmlMultiThemes({
        language: getLanguage(fixture.language),
        themes: formatterThemes,
        defaultTheme: config ? config.defaultTheme : "main",
        highlightLines: config?.highlightLines?.length
          ? { lines: config.highlightLines, style: "theme" }
          : undefined,
      }),
      { rainbowBrackets: fixture.rainbowBrackets },
    );
    expect(output).toBe(fixture.htmlMultiThemes);
  });

  it("bbcodeScoped", () => {
    const output = highlighter.highlight(
      fixture.source,
      bbcodeScoped({
        language: getLanguage(fixture.language),
      }),
      { rainbowBrackets: fixture.rainbowBrackets },
    );
    expect(output).toBe(fixture.bbcode);
  });

  it("terminal", () => {
    const output = highlighter.highlight(
      fixture.source,
      terminal({
        language: getLanguage(fixture.language),
        theme: getTheme(fixture.theme),
      }),
      { rainbowBrackets: fixture.rainbowBrackets },
    );
    expect(output).toBe(fixture.terminal);
  });
});
